core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala

```diff
@@ -254,6 +235,47 @@ class PaigesScalacheckTest extends FunSuite {
     }
   }
 
+  test("Union invariant: `a.flatten == b.flatten`") {
+    forAll { d: Doc.Union => assert(d.a.flatten === d.b.flatten) }
+  }
+
+  test("Union invariant: `a != b`") {
+    forAll { d: Doc.Union => assert(d.a !== d.b) }
+  }
+
+  test("Union invariant: `a` has right-associated `Concat` nodes") {
+    def invariant(d: Doc): Assertion = {
+      import Doc._
+      d match {
+        case Empty | Text(_) | Line(_) => assert(true)
+        case Concat(Concat(_, _), _) => assert(false, s"Left-associative Concat: ${d}")
```

Comment:
what if we allow this, but not `Concat(Concat(Concat(_ ...` I *think* the current fill will pass that. Can we check?

Reply:
Yes, the tests pass.  I'm not confident the `Doc` generator is very thorough, as even removing the `_.grouped` change for the last element let the tests pass, which wouldn't have been the case if it had generated a 3-left-associated `Concat`.

Reply:
I'm wrong about the `_.grouped` comment.  For the invariant to fail, we'd need to generate the first element as a `Union` with the left branch not 2-right-associated.  But assuming `fill` doesn't generate those in the first place, the document generator wouldn't either.  So I think the `grouped` bit is unnecessary; it seems that if the invariant holds in the branches creating the `Union` nodes it will hold for the entire function.  And since it already holds for the only other use of `Union` (in `grouped`) I think we are OK.